### PR TITLE
[3809] Content Tree Panel

### PR DESCRIPTION
**Pull request by aachavarria**
https://github.com/craftercms/craftercms/issues/3809

### CODE REVIEW

File: ui/app/src/components/Icons/Field.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*
```
Comment (rart) [inline ui/app/src/components/Icons/Field.tsx]: Rename to ContentTypeField

File: ui/app/src/components/Icons/Repeat.tsx [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/*
```
Comment (rart) [inline ui/app/src/components/Icons/Repeat.tsx]: Repeat => RepeatGroup

File: ui/app/src/modules/Preview/ToolBar.tsx [hunk 3]
```diff
@@ -290,80 +192,11 @@ export function AddressBar(props: AddressBarProps) {
...
-        dialogConfig.open &&
-        <EmbeddedLegacyEditors dialogConfig={dialogConfig} setDialogConfig={setDialogConfig} getPath={getPath}/>
-      }
-      {
-        publishDialog.open &&
-        <PublishDialog scheduling={publishDialog.scheduling} items={[publishDialog.item]} onClose={onClosePublish}/>
-      }
+        setAnchorEl={setAnchorEl}
+        site={site}
+      />
```
Comment (rart) [inline ui/app/src/modules/Preview/ToolBar.tsx]: - Why does the ComponentMenu need the ability to **set** the anchor?
- Follow same API as material components (use both anchor and open props, onClose, etc)

File: ui/app/src/modules/Preview/Tools/ContentTree.tsx [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+
+import React, { useEffect, useState } from 'react';
+import clsx from 'clsx';
+import { defineMessages, useIntl } from 'react-intl';
+import ToolPanel from './ToolPanel';
+import { createStyles, makeStyles } from '@material-ui/core';
+import TreeView from '@material-ui/lab/TreeView';
+import IconButton from '@material-ui/core/IconButton';
+import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
+import ChevronRightIcon from '@material-ui/icons/ChevronRight';
```
Comment (rart) [inline ui/app/src/modules/Preview/Tools/ContentTree.tsx]: use rounded set

File: ui/app/src/modules/Preview/Tools/ContentTree.tsx [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useEffect, useState } from 'react';
+import clsx from 'clsx';
+import { defineMessages, useIntl } from 'react-intl';
+import ToolPanel from './ToolPanel';
+import { createStyles, makeStyles } from '@material-ui/core';
```
Comment (rart) [inline ui/app/src/modules/Preview/Tools/ContentTree.tsx]: import from `/style`. always avoid importing from `/core`

File: ui/app/src/modules/Preview/Tools/ContentTree.tsx [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+  const hostToGuest$ = getHostToGuestBus();
+  const [expanded, setExpanded] = React.useState<string[]>(['root']);
+  const [data, setData] = React.useState<Data>({
+    previous: [],
+    selected: null,
+    lookupTable: null
+  });
+
+  useEffect(() => {
+    if (guest?.modelId && guest?.models && contentTypesBranch.byId && data.selected === null) {
```
Comment (rart) [inline ui/app/src/modules/Preview/Tools/ContentTree.tsx]: once you've done `guest?.`, on the subsequent uses is not necessary to use the opt chaining op as the first if would have checked for _nullish-ness_

File: ui/app/src/state/actions/preview.ts [hunk 7]
```diff
@@ -64,6 +64,7 @@ export const CONTENT_TYPE_RECEPTACLES_REQUEST = 'CONTENT_TYPE_RECEPTACLES_REQUES
 export const CONTENT_TYPE_RECEPTACLES_RESPONSE = 'CONTENT_TYPE_RECEPTACLES_RESPONSE';
 export const SCROLL_TO_RECEPTACLE = 'SCROLL_TO_RECEPTACLE';
 export const CLEAR_HIGHLIGHTED_RECEPTACLES = 'CLEAR_HIGHLIGHTED_RECEPTACLES';
+export const SCROLL_TO_ELEMENT = 'SCROLL_TO_ELEMENT';
```
Comment (rart) [inline ui/app/src/state/actions/preview.ts]: Since we're also animating, etc use a more specific and semantic action type: CONTENT_TREE_FIELD_SELECTED

File: ui/app/src/styles/icon.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+const iconStyles = {
+  fill: 'none',
+  stroke: 'currentColor'
+};
+
+export default iconStyles;
```
Comment (rart) [inline ui/app/src/styles/icon.ts]: 1. rename const to `iconWithStrokeAndFill`
2. export const
3. don't export default

File: ui/guest/src/classes/ContentController.js [hunk 9]
```diff
@@ -821,6 +821,12 @@ function fetchById(id, site = Cookies.get('crafterSite')) {
                 content_o {
                   ...ContentIncludeWrapperFragment
                 }
+                randomrg_o {
+                  item {
+                    namebonito_s
+                    namefeo_s
+                  }
+                }
```
Comment (rart) [inline ui/guest/src/classes/ContentController.js]: Rollback. Don't commit this.

File: ui/guest/src/components/Guest.js [hunk 10]
```diff
@@ -16,6 +16,8 @@
  */
 
 import React, { useEffect, useRef, useState } from 'react';
+import jss from 'jss';
+import preset from 'jss-preset-default';
```
Comment (rart) [inline ui/guest/src/components/Guest.js]: x

File: ui/guest/src/components/Guest.js [hunk 11]
```diff
@@ -70,6 +73,27 @@ const escape$ = fromEvent(document, 'keydown').pipe(
...
+    'to': {
+      transform: 'scaleX(1)'
+    }
+  },
+  pulseAnim: {
+    animation: '$pulse 300ms 2 ease-in-out'
+  }
+};
+
+const {classes} = jss.createStyleSheet(styles).attach()
```
Comment (rart) [inline ui/guest/src/components/Guest.js]: All these styles go to the styles file. Guest would only apply the class.

File: ui/guest/src/components/Guest.js [hunk 12]
```diff
@@ -45,6 +47,7 @@ import {
   notNullOrUndefined,
   pluckProps,
   RELOAD_REQUEST,
+  SCROLL_TO_ELEMENT,
```
Comment (rart) [inline ui/guest/src/components/Guest.js]: Rename as requested above.

File: ui/guest/src/components/Guest.js [hunk 13]
```diff
@@ -1142,6 +1166,33 @@ export function Guest(props) {
...
+        $(scrollElement).animate({
+          scrollTop: $element.offset().top - 100
+        }, 300, function () {
+          pulseAnimation($element);
+        });
+      } else {
+        pulseAnimation($element);
+      }
+    }
+  }
```
Comment (rart) [inline ui/guest/src/components/Guest.js]: Looks like all these functions can be outside of the component. Move out all that can be moved out - including others below and above not part of this change set.

File: ui/guest/src/util.js [hunk 14]
```diff
@@ -67,6 +67,7 @@ export const CONTENT_TYPE_RECEPTACLES_REQUEST = 'CONTENT_TYPE_RECEPTACLES_REQUES
 export const CONTENT_TYPE_RECEPTACLES_RESPONSE = 'CONTENT_TYPE_RECEPTACLES_RESPONSE';
 export const SCROLL_TO_RECEPTACLE = 'SCROLL_TO_RECEPTACLE';
 export const CLEAR_HIGHLIGHTED_RECEPTACLES = 'CLEAR_HIGHLIGHTED_RECEPTACLES';
+export const SCROLL_TO_ELEMENT = 'SCROLL_TO_ELEMENT';
```
Comment (rart) [inline ui/guest/src/util.js]: Rename as requested above.

**Review by rart — CHANGES_REQUESTED** [review]

File: static-assets/components/cstudio-forms/template-editor.js [hunk 15]
```diff
@@ -274,7 +274,9 @@ CStudioAuthoring.Module.requireModule(
 
                       aceEditor.getSession().on('change', function() {
                         aceEditor.isModified = true;
-                        onSaveCb.pendingChanges();
+                        if (onSaveCb.pendingChanges()) {
```
Comment (rart) [inline static-assets/components/cstudio-forms/template-editor.js]: Did you mean to check for the nullishness or pending changes instead of calling it?

File: ui/app/src/components/ComponentMenu.tsx [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+  useEffect(() => {
+    if (models && modelId && publishDialog.item === null) {
+      let path = models[modelId].craftercms.path;
+      if (embeddedParentPath) path = models[parentId].craftercms.path;
+      getItem(site, path).subscribe(
+        (item) => {
+          setPublishDialog({ item })
+        },
+        (error) => {
+          console.log(error)
```
Comment (rart) [inline ui/app/src/components/ComponentMenu.tsx]: Handle properly.

File: ui/app/src/components/ComponentMenu.tsx [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+        />
+      }
+      {
+        publishDialog.open &&
+        <PublishDialog
+          scheduling={publishDialog.scheduling}
+          items={[publishDialog.item]}
+          onSuccess={onSuccessPublish}
+          onClose={onClosePublish}
+        />
```
Comment (rart) [inline ui/app/src/components/ComponentMenu.tsx:239]: We could set these higher above in the component tree so less components would require this code within them.

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 18]
```diff
@@ -180,7 +183,7 @@ function PublishDialog(props: PublishDialogProps) {
     setOpen(false);
 
     // call externalClose fn
-    onClose();
+    onClose && onClose();
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx]: `onClose?.()`

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 19]
```diff
@@ -203,7 +206,8 @@ function PublishDialog(props: PublishDialogProps) {
       (response) => {
         setOpen(false);
         setApiState({ ...apiState, error: false, submitting: false });
-        onClose(response);
+        onSuccess && onSuccess(response);
+        onClose && onClose(response);
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx]: `onSuccess?.()`
`onClose?.()`

File: ui/app/src/modules/Preview/Tools/ContentTree.tsx [hunk 20]
```diff
@@ -0,0 +1,450 @@
...
+              </>
+            )
+          }
+          <p>{nodes.name}</p>
+          {
+            over && (nodes.type === 'component' || nodes.id === 'root') &&
+            <IconButton className={classes.options} onMouseOver={(e) => setOverState(e, true)}
+                        onClick={(e) => handleOptions(e, nodes.modelId, nodes.parentId, nodes.embeddedParentPath)}>
+              <MoreVertIcon/>
+            </IconButton>
```
Comment (rart) [inline ui/app/src/modules/Preview/Tools/ContentTree.tsx]: Format

**Review by rart — CHANGES_REQUESTED** [review]

File: static-assets/components/cstudio-forms/template-editor.js (hunk 15 above)
Comment (aachavarria) [reply static-assets/components/cstudio-forms/template-editor.js]: nullishness

File: ui/app/src/components/ComponentMenu.tsx (hunk 17 above)
Comment (aachavarria) [reply ui/app/src/components/ComponentMenu.tsx:239]: pending

**Review by rart — APPROVED** [review]

**Review by rart — APPROVED** [review]
